backend/manager/modules/vdsbroker/src/main/java/org/ovirt/engine/core/vdsbroker/VdsManager.java

```diff
@@ -637,20 +637,18 @@ private void handleRefreshCapabilitiesResponse(VDS vds, VDSReturnValue caps) {
         try {
             invokeGetHardwareInfo(vds, caps);
             processRefreshCapabilitiesResponse(new AtomicBoolean(), vds, vds.clone(), caps);
+
+            updateDynamicData(vds.getDynamicData());
+            updateNumaData(vds);
+
+            // Update VDS after testing special hardware capabilities
+            monitoringStrategy.processHardwareCapabilities(vds);
+
+            // Always check VdsVersion
+            resourceManager.getEventListener().handleVdsVersion(vds.getId());
         } catch (Throwable t) {
             logRefreshCapabilitiesFailure(t);
             throw t;
-        } finally {
-            if (vds != null) {
-                updateDynamicData(vds.getDynamicData());
-                updateNumaData(vds);
-
-                // Update VDS after testing special hardware capabilities
-                monitoringStrategy.processHardwareCapabilities(vds);
-
-                // Always check VdsVersion
-                resourceManager.getEventListener().handleVdsVersion(vds.getId());
-            }
```

Comment:
this change seems to improve the current state but still looks wrong - we get here with the result of get-caps, we calls invokeGetHardwareInfo invokes a call to VDSM, that may fail, or process RefreshCapabilitiesResponse that was already received
why would we have an exception here? if getCaps succeeded, it's not likely that invokeGetHardwareInfo would fail (and even if it does, it's not so critical) or the processing of the response of refresh caps fails with exception - that really shouldn't happen

Reply:
I guess caps is null or something like that - do we see that in the log?

Reply:
ah ok, because we check the response in processRefreshCapabilitiesResponse